[onx001] iP

onx001 opened the pull request:
Rebuild Duke as Duchess and add functionality

There are several functions that personal assistants facilitate. Some examples include keeping track of events and deadlines.

Adding all these functionalities help Duchess become more useful.

As a step towards such additions, let's ensure Duchess' code is well organised and sticks to the principles of object-oriented programming for easy function adding. Doing so will make future refactoring and modifications easier.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
Increment lvl-0

Complete week 2 assignments

Rename Duke and complete Greeter exercise
Update gitignore

Commit:
Complete week 2 assignments

Rename Duke and complete Greeter exercise
Update gitignore

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Add Level 1 to coding standards

There are several classes required to handle commands. In some cases, there are inputs and returns.

Let's,
 *Adhere to coding standards for the CS2113 AY23/24 S1 course.
 *Flesh out javadocs.
 *Add functionality and personality to Duchess.

Commit:
Add level 2 functionality

Duchess needs to understand and add to a list of tasks.

Lets,
 *Create a task class to store task specific data
 *Create a task list class to unify functionality of methods such as listing
 *Add more personality to Duchess

Commit:
Change task list index

Humans tend to start listing from 1. Let's change the first number in the list to 1.

Commit:
Add level 3 functionality

Duchess needs a way to understand commands. Some commands target specific tasks and we should identify them.

Let's also separate printing functions in anticipation of different types of tasks.

Commit:
Follow Coding Standard

The instinct is to code for functionality first, rather than readability.

Coding an incremental chatbot to allow for future changes requires standardisation, especially for different developers.

Let's follow the conventions laid out as per se-education.org to remain on the same page and allow for easy understanding and maintaining of code.

This helps not only with future or current differing developers, but also the main developer as code is often forgotten.

Commit:
Refactor code

Current code repeats the continue return condition multiple times.

Let's bring it out of the switch case to avoid repetitive code fragments and improve readability.

Commit:
Add level 4 functionality

There is simple task handling including adding, marking, and unmarking tasks.

Adding more types of tasks will help Duchess become a more useful helper.

Let's,
 *Add To-Do tasks, the simplest form commonly used to indicate something to be done
 *Add deadline tasks, when there is something to be done by a specific time
 *Add events, which happen in a certain timeframe
 *Use inheritance from the Task class to reduce code duplication via centralising common behaviours

Commit:
Add TextUITesting

Automate test cases and edit in "input.txt" file

Typing in test cases repetitively can be tiring.

Let's automate this process with external input and output file checking.

Commit:
Improve A-CodeQuality

Manually setting string positions for commands to be processed can be confusing.

Let's relate it to the constants file to allow for easier future manipulation and understanding.

Commit:
Fix default case

Entering an unknown command should not create a base Task object.

Let's set the default case to print a statement for the user, saying that the command was unrecognised.

aaronxujiachen commented on src/main/java/Duchess/CommandHandler.java:
```diff
@@ -25,13 +25,6 @@ public int ParseCommand(String command) {
-                System.out.println(DefaultStrings.addedString + command);
-                System.out.println(DefaultStrings.splittingLine);
-
             /**Checks for command keywords
```
Clear and succinct comment!

aaronxujiachen commented on src/main/java/Duchess/CommandHandler.java:
```diff
@@ -43,17 +36,54 @@ public int ParseCommand(String command) {
+                System.out.println(DefaultStrings.addedString + command);
+                System.out.println(DefaultStrings.splittingLine);
+                break;
+        } 
```
Good choice to use spacing between cases, looks very neat!

aaronxujiachen commented on src/main/java/Duchess/Task.java:
```diff
@@ -42,12 +50,15 @@ public boolean getDone(){
+            return ("[ ] " + this.name);
         }
     }
 }
```
Indentations and brackets are used properly, good job!

aaronxujiachen commented on src/main/java/Duchess/TaskList.java:
```diff
@@ -18,26 +18,25 @@ public void addTask(Task task){
+                + "\n" + this.tasks.get(taskNumber-1).toString()
+                + "\n" + DefaultStrings.splittingLine);
     }
 
```
Good way to manage printing of long sentences!

Cazh1 commented on src/main/java/Duchess/DefaultStrings.java:
```diff
@@ -0,0 +1,16 @@
+/** Class to define commonly used or long strings for other classes. */
+public class DefaultStrings {
+    public static final String splittingLine = "\t____________________________________________________________\n";
+    public static final String logo = splittingLine +
```
Perhaps a more detailed variable name such as "logoMessage"

Cazh1 commented on src/main/java/Duchess/CommandHandler.java:
```diff
@@ -0,0 +1,87 @@
+    */
+    public int ParseCommand(String command) {
+        String[] commandArray = command.split(" "); // Split input into array of strings
+        switch (commandArray[0]) { // Check first word of input for command
+
+            /**Checks for command keywords
+             * and executes corresponding methods
+             * from TaskList class if found.
+             */
+
+            case Constants.endCommand:
```
Noticed that there was indentation for case clauses.
```suggestion
        switch (commandArray[0]) { // Check first word of input for command

            /**Checks for command keywords
             * and executes corresponding methods
             * from TaskList class if found.
             */

        case Constants.endCommand:
```

Cazh1 reviewed (commented):
Code looks very neat. Noticed some spacing errors

maanyos commented on src/main/java/Duchess/Constants.java:
```diff
@@ -0,0 +1,22 @@
+    public static final String deadlineTime = "/by";
+    public static final String eventStartTime = "/from";
+    public static final String eventEndTime = "/to";
+    public static final int exitFlag = 1;
+    public static final int stayFlag = 0;
```
consider using boolean for 2-state variables?

maanyos commented on src/main/java/Duchess/Event.java:
```diff
@@ -0,0 +1,21 @@
+        super();
+    }
+
+    public Event(String name, String Starttime, String endTime) {
```
Starttime and endTime should have the same casing

maanyos commented on src/main/java/Duchess/Task.java:
```diff
@@ -0,0 +1,64 @@
+    
+    /**Done condition and
+     * Name to be accessed as task */
+    private boolean done = false;
```
follow boolean naming standard

maanyos reviewed (commented):
Interesting decision to use gradle on ip, do prioritise completing the increments on time and that your final jar file is runnable without gradle.
Good decision to start documenting early, should make things easier later on.
Generally good adherence to OOP principles and coding standards, with a few minor exceptions here and there

onx001 committed:
Add increment Level 5 to separate branch

Users might input incomplete tasks, which need to be filtered.

Let's add error handling to the addition of tasks.

onx001 committed:
Revert "Add increment Level 5 to separate branch"

This reverts commit a99698c143d266331a47e1f53b3bfc66049a4a59.

onx001 committed:
Revert "Revert "Add increment Level 5 to separate branch""

This reverts commit 1fc82f59b4d30f3dc5e4c33ebc5270172cceee48.

onx001 committed:
Revert "Revert "Revert "Add increment Level 5 to separate branch"""

This reverts commit ab9d484d44129826abbb1138a3911973c6771548.

onx001 committed:
Update Duchess.java

Add commit for merging exercise

onx001 committed:
Merge branch 'branch-Level-5'

onx001 committed:
Organize into packages

Different class objects have different purposes, some of which can be grouped.

Let's formalise the groupings by introducing 3 packages.

onx001 committed:
Merge branch 'branch-A-Packages'

onx001 committed:
Revert "Revert "Revert "Revert "Add increment Level 5 to separate branch""""

This reverts commit 27d8481d274f4689b400915bf5463a9e925100f2.

onx001 committed:
Use Exceptions to handle errors

Printing error handling messages by the command handler can be messy.

Let's create a separate error handling mechanism to help make complex future tasks easier to manage individually.

onx001 committed:
Refactor for string handler

In the command handler class, there are multiple instances of string manipulation.

Let's collate these instances and refactor them into a new class called StringHandler.

onx001 committed:
Add support for deleting tasks from list

Sometimes, the user adds tasks they do not intend to do.

Let's add a "delete task" function to allow for users who wish to streamline their lists.

onx001 committed:
Use Java Collections classes

Highlight areas where java collections classes is being used.

onx001 committed:
Add save fuction

Users might want to save tasks in the hard disk automatically and reuse the program over a span of time.

Let's read and write the tasks for long term storage so that the program can be used in separate sessions without losing data.

onx001 committed:
Handle non-existence case

The file might not be created by new users manually.

Let's create a new file for them automatically.

onx001 committed:
Merge branch 'branch-Level-6'

onx001 committed:
Merge branch 'branch-Level-7'

onx001 committed:
Edit git ignore for jar file

Binaries are not typically committed in git projects.

Let's exclude the file.

onx001 committed:
A-MoreOOP

Consolidation of functions into classes is useful for debugging and adding new features.

Let's add a UI class to handle most of the print statements.

In the future, let's also move UI functions from the TaskList class into the new UI class.

onx001 committed:
Add date and time functionality

Deadlines and events sometimes need to be operated on as numerical quantities.

Let's add date and time functionality to Duchess tasks for additional future features.

onx001 committed:
Add find functionality

Searching for related tasks is important for any tracker, especially when there are many tasks.

Let's add a search function to allow users to look for tasks.

onx001 committed:
A-JavaDoc

Class and function descriptors are important for quick and easy reference.

Let's add javadocs to public methods and classes for easy further development.

onx001 committed:
Merge pull request #1 from onx001/branch-Level-8

Add date and time functionality

onx001 committed:
Merge branch 'master' into branch-Level-9

onx001 committed:
Merge pull request #2 from onx001/branch-Level-9

Add find functionality

onx001 committed:
Merge branch 'master' into branch-A-JavaDoc

onx001 committed:
Merge pull request #3 from onx001/branch-A-JavaDoc

A-JavaDoc

onx001 committed:
Start work on user guide

onx001 committed:
Update README.md

onx001 committed:
Update User Guide

onx001 committed:
A-UserGuide

User guides are helpful for end users to navigate the program.

Let's create a useful user guide to accompany Duchess.

onx001 committed:
Update README.md

onx001 committed:
Update README.md

Let's correct minor formatting errors.

onx001 committed:
Merge branch 'master' of https://github.com/onx001/ip

onx001 committed:
Fix minor bugs
